### PR TITLE
[FLINK-27185][connectors] Convert connector-jdbc module to assertj

### PR DESCRIPTION
**Pull request by alpreu**
Cherry-picked commits from https://github.com/apache/flink/pull/19660 to get the individual changes merged faster

### CODE REVIEW

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/PostgresCatalogITCase.java [hunk 1]
```diff
@@ -138,9 +138,9 @@ public void testPrimitiveTypes() {
                                 .execute()
                                 .collect());
 
-        assertEquals(
-                "[+I[1, [50], 3, 4, 5.5, 6.6, 7.70000, 8.8, true, a, b, c  , d, 2016-06-22T19:10:25, 2015-01-01, 00:51:03, 500.000000000000000000]]",
-                results.toString());
+        assertThat(results.toString())
+                .isEqualTo(
+                        "[+I[1, [50], 3, 4, 5.5, 6.6, 7.70000, 8.8, true, a, b, c  , d, 2016-06-22T19:10:25, 2015-01-01, 00:51:03, 500.000000000000000000]]");
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/PostgresCatalogITCase.java:143]:
```suggestion
        assertThat(results).hasToString(
                "[+I[1, [50], 3, 4, 5.5, 6.6, 7.70000, 8.8, true, a, b, c  , d, 2016-06-22T19:10:25, 2015-01-01, 00:51:03, 500.000000000000000000]]");
```

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/PostgresCatalogITCase.java [hunk 2]
```diff
@@ -151,28 +151,28 @@ public void testArrayTypes() {
...
+                                + "[true, false, true], "
+                                + "[a, b, c], "
+                                + "[b, c, d], "
+                                + "[b  , c  , d  ], "
+                                + "[b, c, d], "
+                                + "[2016-06-22T19:10:25, 2019-06-22T19:10:25], "
+                                + "[2015-01-01, 2020-01-01], "
+                                + "[00:51:03, 00:59:03], "
+                                + "null, "
+                                + "null]]");
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/PostgresCatalogITCase.java:175]: hasToString

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/PostgresCatalogITCase.java [hunk 3]
```diff
@@ -183,14 +183,14 @@ public void testSerialTypes() {
...
-                results.toString());
+        assertThat(results.toString())
+                .isEqualTo(
+                        "[+I["
+                                + "32767, "
+                                + "2147483647, "
+                                + "32767, "
+                                + "2147483647, "
+                                + "9223372036854775807, "
+                                + "9223372036854775807]]");
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/catalog/PostgresCatalogITCase.java:194]: hasToString

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/dialect/oracle/OracleTableSourceITCase.java [hunk 4]
```diff
@@ -240,9 +239,9 @@ public void testLimit() throws Exception {
...
         expected.add(
                 "+I[2, 2020-01-01T15:36:01.123456, 2020-01-01T15:36:01.123456789, -1.175E-10, -1.79769E40, 101.1234]");
-        assertEquals(1, result.size());
-        assertTrue(
-                "The actual output is not a subset of the expected set.",
-                expected.containsAll(result));
+        assertThat(result).hasSize(1);
+        assertThat(expected.containsAll(result))
+                .as("The actual output is not a subset of the expected set.")
+                .isTrue();
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/dialect/oracle/OracleTableSourceITCase.java:245]:
```suggestion
        assertThat(expected)
                .as("The actual output is not a subset of the expected set.").containsAll(result);
```

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/internal/connection/SimpleJdbcConnectionProviderTest.java [hunk 5]
```diff
@@ -71,49 +61,49 @@ private static JdbcConnectionProvider newFakeConnectionProvider() {
...
-        assertNotNull(provider.getConnection());
-        assertSame(connection, provider.getConnection());
-        assertSame(connection, provider.getOrEstablishConnection());
+        assertThat(connection).isNotNull();
+        assertThat(connection.isClosed()).isFalse();
+        assertThat(provider.isConnectionValid()).isTrue();
+        assertThat(connection).isInstanceOf(FakeConnection.class);
+
+        assertThat(provider.getConnection()).isNotNull();
+        assertThat(provider.getConnection()).isSameAs(connection);
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/internal/connection/SimpleJdbcConnectionProviderTest.java:74]: could be joined as
```suggestion
        assertThat(provider.getConnection()).isNotNull()
                .isSameAs(connection);
```

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/internal/connection/SimpleJdbcConnectionProviderTest.java [hunk 6]
```diff
@@ -71,49 +61,49 @@ private static JdbcConnectionProvider newFakeConnectionProvider() {
...
-        assertSame(connection, provider.getConnection());
-        assertSame(connection, provider.getOrEstablishConnection());
+        assertThat(connection).isNotNull();
+        assertThat(connection.isClosed()).isFalse();
+        assertThat(provider.isConnectionValid()).isTrue();
+        assertThat(connection).isInstanceOf(FakeConnection.class);
+        assertThat(connection).isNotInstanceOf(FakeConnection3.class);
+
+        assertThat(provider.getConnection()).isNotNull();
+        assertThat(provider.getConnection()).isSameAs(connection);
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/internal/connection/SimpleJdbcConnectionProviderTest.java:92]: could be joined as
```suggestion
        assertThat(connection).isInstanceOf(FakeConnection.class)
                .isNotInstanceOf(FakeConnection3.class);

        assertThat(provider.getConnection()).isNotNull()
                .isSameAs(connection);
```

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcRowDataInputFormatTest.java [hunk 7]
```diff
@@ -234,7 +234,7 @@ public void testJdbcInputFormatWithParallelismAndNumericColumnSplitting() throws
         inputFormat.openInputFormat();
         InputSplit[] splits = inputFormat.createInputSplits(1);
         // this query exploit parallelism (1 split for every id)
-        Assert.assertEquals(TEST_DATA.length, splits.length);
+        assertThat(splits).hasSize(TEST_DATA.length);
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcRowDataInputFormatTest.java:237]:
```suggestion
        assertThat(splits).hasSameSizeAs(TEST_DATA);
```

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcRowDataInputFormatTest.java [hunk 8]
```diff
@@ -311,7 +311,7 @@ public void testJdbcInputFormatWithParallelismAndGenericSplitting() throws IOExc
         inputFormat.openInputFormat();
         InputSplit[] splits = inputFormat.createInputSplits(1);
         // this query exploit parallelism (1 split for every queryParameters row)
-        Assert.assertEquals(queryParameters.length, splits.length);
+        assertThat(splits).hasSize(queryParameters.length);
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcRowDataInputFormatTest.java:314]:
```suggestion
        assertThat(splits).hasSameSizeAs(queryParameters);
```

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/xa/XidImplTest.java [hunk 9]
```diff
@@ -35,24 +34,26 @@ public void testXidsEqual() {
         XidImpl other =
                 new XidImpl(
                         XID.getFormatId(), XID.getGlobalTransactionId(), XID.getBranchQualifier());
-        assertEquals(XID, other);
-        assertEquals(XID.hashCode(), other.hashCode());
+        assertThat(other).isEqualTo(XID);
+        assertThat(other.hashCode()).isEqualTo(XID.hashCode());
```
Comment (snuyanzin) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/xa/XidImplTest.java:38]:
```suggestion
        assertThat(other).isEqualTo(XID)
                .hasSameHashCodeAs(XID);
```

**Review by snuyanzin — COMMENTED** [review]
In general lgtm, I left some comments

**Comment by alpreu** [comment]
Thanks for taking the time to review @snuyanzin, I just pushed again with the requested changes

**Comment by snuyanzin** [comment]
LGTM as soon as the build completes successfully

**Comment by afedulov** [comment]
LGTM

**Comment by XComp** [comment]
@alpreu Sorry for replying that late, but what about also migrating from junit4 to junit5 as part of that effort? Since we decided to go for assertj/junit5 in general? What's the motivation of moving to assertj only?

**Comment by alpreu** [comment]
> @alpreu Sorry for replying that late, but what about also migrating from junit4 to junit5 as part of that effort? Since we decided to go for assertj/junit5 in general? What's the motivation of moving to assertj only?

The migration to assertJ was mainly done using the script written by @slinkydeveloper (https://github.com/assertj/assertj-migrator). We did have to manually touch some files though because not all assertions worked out of the box and some had better alternatives. I believe there is another umbrella ticket tracking the Junit5 migration

**Review by XComp — COMMENTED** [review]
Thanks, @alpreu . In the end, I decided to go over the diff once more as well because I was curious about the assertj migration. I identified a few nitty things which we might want to fix before merging...

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcOutputFormatTest.java [hunk 10]
```diff
@@ -91,181 +87,190 @@ public void tearDown() throws Exception {
...
+                                    new JdbcOutputFormatBuilder()
+                                            .setJdbcOptions(jdbcOptions)
+                                            .setFieldDataTypes(fieldDataTypes)
+                                            .setJdbcDmlOptions(dmlOptions)
+                                            .setJdbcExecutionOptions(
+                                                    JdbcExecutionOptions.builder().build())
+                                            .build();
+                            outputFormat.open(0, 1);
+                        })
+                .hasMessage(expectedMsg);
```
Comment (XComp) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcOutputFormatTest.java:115]: We're not checking for the `IOException` here 🤔

**Review by XComp — COMMENTED** [review]
Thanks for addressing my comments. It looks good now. Two question popped up, though, when doing my final pass over the change. Please see them below...

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/JdbcRowOutputFormatTest.java [hunk 11]
```diff
@@ -76,27 +75,23 @@ public void testInvalidDriver() {
...
-        }
+
+        jdbcOutputFormat =
+                JdbcRowOutputFormat.buildJdbcOutputFormat()
+                        .setDrivername(DERBY_EBOOKSHOP_DB.getDriverClass())
+                        .setDBUrl("jdbc:der:iamanerror:mory:ebookshop")
+                        .setQuery(String.format(INSERT_TEMPLATE, INPUT_TABLE))
+                        .finish();
+        assertThatThrownBy(() -> jdbcOutputFormat.open(0, 1))
+                .satisfies(anyCauseMatches(SQLException.class, expectedMsg));
```
[the remark this answers is not in the document]
Comment (alpreu) [reply flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/JdbcRowOutputFormatTest.java:94]: The changed method has a different check for the class+message of the throwable, I believe previously it checked for the first throwables class with any method down the stacktrace. Now it checks that in the stacktrace any throwable has the given class and message

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcOutputFormatTest.java (hunk 10 above)
Comment (alpreu) [reply flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcOutputFormatTest.java:115]: Good catch :)

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/JdbcRowOutputFormatTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (XComp) [reply flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/JdbcRowOutputFormatTest.java:94]: But shouldn't we check for the `IOException` as well?

[the remark this answers is not in the document]
Comment (XComp) [reply flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/JdbcRowOutputFormatTest.java:94]: I verified that the stacktrace is, indeed, how you described it with `IOException` being caused by a `SQLException`. But I don't know enough about the use-case and the intention of this test. Therefore, I would assume that we should do the conversion properly by also asserting the `IOException` here...

**Review by XComp — APPROVED** [review]
LGTM 👍

**Comment by XComp** [comment]
The CI failure isn't related to the change but is caused by FLINK-27791. I'm going to go ahead and merge the PR.

**Pull request merged**